flatbuffers/flatbuffers_case_study.md

```diff
@@ -0,0 +1,172 @@
+console.log('event data', event.event_data(new EventSerializer.MousePositionData())).x()
+```
+
+As you can see, Flatbuffers creation and reading is not trivial. The API is verbal and cumbersome.
+
+# Phases of refactor
+Refactoring our entire flow from JSON to FlatBuffers is quite a feat. It's a architecture crossing change that will be dangerous to try and accomplish in one go. We've considered this issue for a while and managed to split the redactor into several independent steps.
+
+## FlatBuffers in recorder
+We began in transforming our recorder messages to FlatBuffers. The recorder creates and send the recorded events as FlatBuffers. The first step of the processing pipeline is to transform the binary data back to JSON allowing the rest of the architecture to stay unchanged.
+
+There is good value to this seemingly small change. It would allow us to skip the process blocking JSON.stringify, and minimize the amount of data the end user needs to upload to our server.
+
+## FlatBuffers in player
+The second step we took was to add FlatBuffers support to our player. Naturally we needed to convert the JSON data we store to FlatBuffers but this extra converting code is a small price to pay in order to factorize a big refactor.
+
+## Store FlatBuffers - "point of no return"
+The Final step is to actually store the FlatBuffers bytes in our storage. This is the "point of no return". In previous steps we could easily roll back the change. Once we loose JSON all together, a rollback would be very difficult.
+
+In fact we have not taken this step yet. A Major benefit of JSON is that it's easily streamable using libraries such as [Oboe.js](http://oboejs.com/). We have yet to discover a method to achieve a similar goal using FlatBuffers.
+
+# benchmarks
+## Space banchmark
+We have expected FlatBuffers to reduce the size of our data. We've converted some of our JSON sessions to FlatBuffers to check this hypothesis. Each tested session is from a different customer, and has different characteristics, so variation in the results are expected.
+
+<scatter plot>
+
+The result did not met our expectations, not only the FlatBuffers file are not smaller that our JSON, they are 15%-30% bigger. While this is a disappointing result the overall file size is just a secondary goal. This issue will definitely be studies.
+
+## Data availability speed benchmark
+The next important benchmark would be data availability speed. By this I mean the duration of time it takes from initial request to complete data availability. In this benchmark we will take several session, in both JSON and FlatBuffers format, download and iterate over their data to perform some simple calculation.
+
+<some kind of graph>
+
+As expected, JSON is impeded by the parsing stage. FlatBuffers allows immediate access to data after download. Those several seconds of JSON parsing are critical for us. These are the first seconds of our page loads, where the first impression of our system takes place.
```

Comment:
`These are the first seconds of our page loads, where the first impression of our system takes place.`
something is off with this sentence